### PR TITLE
KAA-290: Added page with information about device provisioning and endpoint registration.

### PR DESCRIPTION
**Pull request by v-chyzhevskyi**
Added page with information about device provisioning and endpoint registration.

### CODE REVIEW

File: doc/Programming-guide/Key-platform-features/Devices-provisioning-and-registration/index.md [hunk 1]
```diff
@@ -6,5 +6,131 @@ nav: /:path/Programming-guide/Key-platform-features/Devices-provisioning-and-reg
...
+Kaa provides two credentials service
+
+1. Trustful credentials service -  allows any endpoint to register and connect to Kaa cluster. (like the previous version of Kaa)
+2. Internal credentials service -  allows connecting with Kaa for specified list of endpoints whose credentials was previously provisioned with Kaa REST API.
+
+When Tenant Admin create application he can specify credential service for this application. By default used "Trustful".
+
+## Creating custom credentials service
+
+1. First of all you should create class which implements all method of CredentialsService interface
```
Comment (ashvayka) [inline doc/Programming-guide/Key-platform-features/Devices-provisioning-and-registration/index.md]: which implements CredentialsService(link to github file in master branch) interface

File: doc/Programming-guide/Key-platform-features/Devices-provisioning-and-registration/index.md [hunk 2]
```diff
@@ -6,5 +6,131 @@ nav: /:path/Programming-guide/Key-platform-features/Devices-provisioning-and-reg
...
+Build your application using next command: 
+
+```bash
+
+$ mvn clean install
+
+```
+
+Place your application *.jar from /target folder into the /usr/lib/kaa-node/lib folder.
+Specify additional package to scan kaa plugins configuration in parameter additional_plugins_scan_package.
```
Comment (ashvayka) [inline doc/Programming-guide/Key-platform-features/Devices-provisioning-and-registration/index.md]: please specify some command how to edit configuration file ( no name of configuration file mentioned here)

File: doc/Programming-guide/Key-platform-features/Devices-provisioning-and-registration/index.md [hunk 3]
```diff
@@ -6,5 +6,131 @@ nav: /:path/Programming-guide/Key-platform-features/Devices-provisioning-and-reg
...
 ---
 
-Devices provisioning and registration
+* TOC
+{:toc}
 
+{% assign root_url = page.url | split: '/'%}
+{% capture root_url  %} /{{root_url[1]}}/{{root_url[2]}}/{% endcapture %}
+
+Kaa establishes a mechanism to uniquely identify each endpoint and the associated tokens and credentials issued to that endpoint.
```
Comment (ashvayka) [inline doc/Programming-guide/Key-platform-features/Devices-provisioning-and-registration/index.md:15]: Please add a section "## Endpoint Credentials" and move all Public/Private keys stuff there. Add a note that this is a default implementation and one can customize it using custom transport implementation.

Comment (ashvayka) [inline doc/Programming-guide/Key-platform-features/Devices-provisioning-and-registration/index.md:15]: In the first section you should talk about overall credentials service functionality (provision, revoke, mark in use).

File: doc/Programming-guide/Key-platform-features/Devices-provisioning-and-registration/index.md [hunk 4]
```diff
@@ -6,5 +6,131 @@ nav: /:path/Programming-guide/Key-platform-features/Devices-provisioning-and-reg
...
+
+Credential is an object that is verified when presented to the verifier in an authentication transaction. 
+It establish the identity of endpoint. 
+
+Kaa provides two credentials service
+
+1. Trustful credentials service -  allows any endpoint to register and connect to Kaa cluster. (like the previous version of Kaa)
+2. Internal credentials service -  allows connecting with Kaa for specified list of endpoints whose credentials was previously provisioned with Kaa REST API.
+
+When Tenant Admin create application he can specify credential service for this application. By default used "Trustful".
```
Comment (balldir) [inline doc/Programming-guide/Key-platform-features/Devices-provisioning-and-registration/index.md]: > When Tenant Admin create**s** application he can specify credential service for this application. By default **"Trustful" is used**.

File: doc/Programming-guide/Key-platform-features/Devices-provisioning-and-registration/index.md (hunk 1 above)
Comment (balldir) [inline doc/Programming-guide/Key-platform-features/Devices-provisioning-and-registration/index.md]: Is "First of all" really necessary ? It is already numbered as 1.

File: doc/Programming-guide/Key-platform-features/Devices-provisioning-and-registration/index.md [hunk 5]
```diff
@@ -6,5 +6,131 @@ nav: /:path/Programming-guide/Key-platform-features/Devices-provisioning-and-reg
...
+<bean id="customCredentialsService" class="org.myproject.CustomCredentialsService"/>
+<alias name="customCredentialsService" alias="NEW Credentials Service"/>
+
+```
+
+##  Custom credentials service provisioning
+
+To provision your credentials service, do the following:
+
+Build your application using next command:
```
Comment (balldir) [inline doc/Programming-guide/Key-platform-features/Devices-provisioning-and-registration/index.md]: Should it be a numbered list that describes steps one-by-one?

File: doc/Programming-guide/Key-platform-features/Devices-provisioning-and-registration/index.md [hunk 6]
```diff
@@ -6,5 +6,131 @@ nav: /:path/Programming-guide/Key-platform-features/Devices-provisioning-and-reg
...
+For this use [Admin REST API]({{root_url}}Programming-guide/Server-REST-APIs/#TODO).
+Registration of public keys occurs only once. In such an approach, all message will be encrypted with a recipient's public key. 
+The message cannot be decrypted by anyone who does not possess the matching private key, who is thus presumed to be the owner of that key and the endpoint associated with the public key. 
+Also you can disable the access to server for endpoint by terminate all active sessions with this endpoint. All subsequent requests will be forbidden.
+For this use [Admin REST API]({{root_url}}Programming-guide/Server-REST-APIs/#TODO).
+
+## Credentials service
+
+Credential is an object that is verified when presented to the verifier in an authentication transaction. 
+It establish the identity of endpoint.
```
Comment (balldir) [inline doc/Programming-guide/Key-platform-features/Devices-provisioning-and-registration/index.md:20]: > It establish**es** the identity of endpoint

**Comment by abohomol** [comment]
I'd mark this as extremely important doc, please review/merge.

**Comment by v-chyzhevskyi** [comment]
This pull request is deprecated.
Please review this new pull request 
https://github.com/kaaproject/kaa/pull/899

**Pull request closed without merging**